### ISSUE TITLE
Projucer menu item "Close all documents" does not always do what it says

### ISSUE THREAD

**Issue by KoenTanghe** [WeAreROLI/JUCE#103]
1) detailed steps on how to reproduce the bug, preferably with already
     existing JUCE code such as the demo plugin or the demo code

Open any .jucer project, and open lots of source files of the project in the Projucer (like 20 or more).
Then click the menu item "Close all documents".
Sometimes a bunch of files will get closed leaving many still open, sometimes only 1 file will close, and sometimes all will close. I didn't notice a pattern as to when this goes wrong...
Note that this happens even when you don't make any changes to the files, and after closing them all manually, they are still unmodified (so it's not that the Projucer modifies some of them and it keeps them from being closed or something).

  2) the expected behaviour

All open documents should be closed (and for all files that were changed, a Save dialog box should be shown, but this is irrelevant for this bug I think as it also happens when nothing was changed).
Post-condition: no document should still be open after this command.

  3) the operating system

Windows 7 64-bit, English, latest service packs and updates.

  4) the architecture (32-bit or 64-bit)

64-bit release version of the Projucer.

  5) a stack trace - if the bug causes a crash

(not a crashing bug)

It's there for a long time already, but to make sure it's there with the latest versions, I just pulled the Juce develop branch, rebuilt the Projucer and also tried in that very last version: same problem. 

It's not critical, but rather anoying if you have lots of files open.

**Comment by tpoole** [comment]
I'm afraid that we just can't reproduce this behaviour.

I've tried opening and closing all 46 files in the JuceDemo example on a few different machines. Can you think of anything else that might help us track this down?

**Comment by KoenTanghe** [comment]
Sorry for the delay.

Hmmm... Could it be that it has something to do with having GUI components among the code files for which the GUI editor was/is used to create them (with Add new GUI component...)?
IIRC, the Juce demo project doesn't have that, while our projects do.

Is there an example project in the Juce repository that has .cpp files which were generated using the GUI editor in the Projucer (our own files got created using the GUI designer in the older Introjucer still, but got opened and saved in the latest Projucer)?

**Comment by tpoole** [comment]
Is this still present on the latest version? I still can't reproduce the issue.

**Comment by KoenTanghe** [comment]
Still present on the develop branch (freshly just pulled just a few minutes ago).
I would open the project, click on some .cpp files for GUI components in the sidebar tree with the list of files, and then go to the "subcomponents" tab to see the GUI layout for the component.
I usually end up with multiple .cpp component files open, and then at some point I go to Windows --> Close all windows, and Projucer notices that some files were changed, ask to save them, and then after asking me for saving (which I confirm), it would stop doing that. But still many files remain open.
If I then do that again, it will again say some files (different ones, still open) have changed, asks for saving etc...
It's as if the loop that goes through all open files that have changed, is broken after saving some of them. Then if you it again, it will continue with the ones that it couldn't handle the previous time you chose "Close all windows".

Does the Projucer have a log file I could send you? Maybe something goes wrong while saving one of the changed files, and that gets logged in a log file somewhere?

As said, no big issue, just annoying.

**Comment by ed95** [comment]
Should be fixed with commit [f961456](https://github.com/WeAreROLI/JUCE/commit/f96145682a86c2c3a7be2530dee20a47e452d79b)

**Comment by KoenTanghe** [comment]
Thanks for looking into this.

I pulled the latest Juce develop branch, built the ProJucer, ran it and then re-opened my project.
Some component source files were still open (Projucer probably keeps some history somewhere?), and when I did Close all windows, it asked to save them. For all files I declined saving, and then all of a sudden, Projucer crashed (while I was clicking the dialog box button to decline saving).

I recompiled Projucer in Debug mode and did the previous steps again, so I would get a call stack when the crash would happen again, and it did: see attached screenshot of the call stack + note the "dead" memory in the variables panel.

![crashing_projucer](https://user-images.githubusercontent.com/872038/29636951-cc74e3bc-8852-11e7-82dd-06c591b396be.png)
